presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/GSheetsDataProvider.java

```diff
@@ -0,0 +1,205 @@
+import com.google.api.services.sheets.v4.model.ValueRange;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.LoadingCache;
+import io.airlift.log.Logger;
+import io.prestosql.spi.PrestoException;
+import org.apache.http.annotation.ThreadSafe;
+
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.OptionalLong;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import static com.google.common.cache.CacheLoader.asyncReloading;
+import static com.google.common.cache.CacheLoader.from;
+import static com.google.common.util.concurrent.MoreExecutors.newDirectExecutorService;
+import static io.prestosql.plugin.google.sheets.SheetsErrorCode.SHEETS_BAD_CREDENTIALS_ERROR;
+import static io.prestosql.plugin.google.sheets.SheetsErrorCode.SHEETS_METASTORE_ERROR;
+import static io.prestosql.plugin.google.sheets.SheetsErrorCode.SHEETS_UNKNOWN_TABLE_ERROR;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+@ThreadSafe
+public class GSheetsDataProvider
+        implements SheetsDataProvider
+{
+    private static final Logger log = Logger.get(GSheetsDataProvider.class);
+    private static final String APPLICATION_NAME = "presto google sheets integration";
+    private static final JsonFactory JSON_FACTORY = JacksonFactory.getDefaultInstance();
+
+    private static final List<String> SCOPES =
+            Collections.singletonList(SheetsScopes.SPREADSHEETS_READONLY);
+
+    private final SheetsConfig sheetsConfig;
+    // tableName <-> sheetId mapping
+    private final LoadingCache<String, Optional<String>> tableSheetMappingCache;
+
+    // sheetId <-> Row List mapping
+    private final LoadingCache<String, Optional<List<List<Object>>>> sheetDataCache;
+
+    private Credential credential;
+    private NetHttpTransport httpTransport;
+    private Sheets sheetsService;
+
+    public GSheetsDataProvider(SheetsConfig config)
+    {
+        this.sheetsConfig = config;
+
+        try {
+            this.credential = getCredentials();
+            this.httpTransport = GoogleNetHttpTransport.newTrustedTransport();
+            this.sheetsService = new Sheets.Builder(httpTransport, JSON_FACTORY, this.credential).setApplicationName(APPLICATION_NAME).build();
+        }
+        catch (Exception e) {
+            throw new PrestoException(SHEETS_BAD_CREDENTIALS_ERROR, e);
+        }
+        OptionalLong expiresAfterWriteMillis = OptionalLong.of(TimeUnit.MINUTES.toMillis(sheetsConfig.getSheetsDataExpireAfterWriteMinutes()));
+        OptionalLong refreshMillis = OptionalLong.of(TimeUnit.MINUTES.toMillis(sheetsConfig.getSheetsDataExpireAfterAccessMinutes()));
+        long maxCacheSize = 1000;
+        ExecutorService executor = newDirectExecutorService();
```

Comment:
does refreshing work with direct executor? Maybe let's remove the refreshing option.

Reply:
good point, removing refreshing options.